utils/utils.c

```diff
@@ -845,3 +849,64 @@ int AlmostEqualRelativeAndAbs_double(double A, double B,
 }
 
 /* #undef __USE_XOPEN2K */
+
+/* A parallel cumulative sum
+   Output convention is: cumsum[0] = 0; cumsum[N-1] = sum(a[0:N-1]);
+   The algorithm is:
+   - Divide the array into `nthreads` chunks
+   - cumsum within each chunk
+   - compute the "offset" for each chunk by summing the cumsum at the tail of all previous chunks
+   - apply the offset
+*/
+void parallel_cumsum(const int64_t *a, const int64_t N, int64_t *cumsum){
+    #ifdef _OPENMP
+    int nthreads = omp_get_max_threads();
+    #else
+    int nthreads = 1;
+    #endif
+    
+    int64_t min_N_per_thread = 10000;  // heuristic to avoid overheads
+    if(N/min_N_per_thread < nthreads){
+        nthreads = N/min_N_per_thread;
+    }
+    if(nthreads < 1){
+        nthreads = 1;
+    }
+    
+    #ifdef _OPENMP
+    #pragma omp parallel num_threads(nthreads)
+    #endif
+    {
+        #ifdef _OPENMP
+        int tid = omp_get_thread_num();
+        #else
+        int tid = 0;
+        #endif
+        
+        int64_t cstart = N*tid/nthreads;
+        int64_t cend = N*(tid+1)/nthreads;
```

Comment:
Does the code still work if `N < nthreads`?

Reply:
It's safe because prior code limits `nthreads <= N`.  I've added an explanatory comment, and a protection for the degenerate case `N=0`, which actually would have resulted in undefined behavior.